truffle/src/main/java/org/jruby/truffle/runtime/object/ObjectGraph.java

```diff
@@ -74,13 +79,26 @@ public void run(DynamicObject thread, Node currentNode) {
                     }
 
                     try {
-                        // We only visit the global variables from the root thread
-                        visitObject(context.getCoreLibrary().getGlobalVariablesObject(), visitor);
+                        // We only visit the global variables and other global state from the root thread
+
+                        if (thread == context.getThreadManager().getRootThread()) {
```

Comment:
This will run for all fibers of the main thread. One way to only run for the initial Fiber is to compare with the root fiber of the main thread.